lib/parachain/types.go

```diff
@@ -0,0 +1,395 @@
+	// The descriptor of the candidate occupying the core.
+	CandidateDescriptor CandidateDescriptor `scale:"8"`
+}
+
+// Index returns the index
+func (OccupiedCore) Index() uint {
+	return 0
+}
+
+// ScheduledCore Information about a core which is currently occupied.
+type ScheduledCore struct {
+	// The ID of a para scheduled.
+	ParaID ParaID `scale:"1"`
+	// The collator required to author the block, if any.
+	Collator *CollatorID `scale:"2"`
+}
+
+// Index returns the index
+func (ScheduledCore) Index() uint {
+	return 1
+}
+
+// Free Core information about a core which is currently free.
+type Free struct{}
+
+// Index returns the index
+func (Free) Index() uint {
+	return 2
+}
+
+// CoreState represents the state of a particular availability core.
+type CoreState scale.VaryingDataType
+
+// Set will set a VaryingDataTypeValue using the underlying VaryingDataType
+func (va *CoreState) Set(val scale.VaryingDataTypeValue) (err error) {
+	// cast to VaryingDataType to use VaryingDataType.Set method
+	vdt := scale.VaryingDataType(*va)
+	err = vdt.Set(val)
+	if err != nil {
+		return fmt.Errorf("setting value to varying data type: %w", err)
+	}
+	// store original ParentVDT with VaryingDataType that has been set
+	*va = CoreState(vdt)
+	return nil
+}
+
+// Value returns the value from the underlying VaryingDataType
+func (va *CoreState) Value() (scale.VaryingDataTypeValue, error) {
+	vdt := scale.VaryingDataType(*va)
+	return vdt.Value()
+}
+
+// NewCoreStateVDT returns a new CoreState VaryingDataType
+func NewCoreStateVDT() (scale.VaryingDataType, error) {
+	vdt, err := scale.NewVaryingDataType(OccupiedCore{}, ScheduledCore{}, Free{})
+	if err != nil {
+		return scale.VaryingDataType{}, fmt.Errorf("create varying data type: %w", err)
+	}
+
+	return vdt, nil
+}
+
+// NewAvailabilityCores returns a new AvailabilityCores
+func NewAvailabilityCores() (scale.VaryingDataTypeSlice, error) {
+	vdt, err := NewCoreStateVDT()
+	if err != nil {
+		return scale.VaryingDataTypeSlice{}, fmt.Errorf("create varying data type: %w", err)
+	}
+
+	return scale.NewVaryingDataTypeSlice(vdt), nil
+}
+
+// UpwardMessage A message from a parachain to its Relay Chain.
+type UpwardMessage []byte
+
+// OutboundHrmpMessage is an HRMP message seen from the perspective of a sender.
+type OutboundHrmpMessage struct {
+	Recipient uint32 `scale:"1"`
+	Data      []byte `scale:"2"`
+}
+
+// ValidationCode is Parachain validation code.
+type ValidationCode []byte
+
+// headData is Parachain head data included in the chain.
+type headData []byte
+
+// CandidateCommitments are Commitments made in a `CandidateReceipt`. Many of these are outputs of validation.
+type CandidateCommitments struct {
+	// Messages destined to be interpreted by the Relay chain itself.
+	UpwardMessages []UpwardMessage `scale:"1"`
+	// Horizontal messages sent by the parachain.
+	HorizontalMessages []OutboundHrmpMessage `scale:"2"`
+	// New validation code.
+	NewValidationCode ValidationCode `scale:"3"`
```

Comment:
the type must be a pointer to `ValidationCode` as it's  `Option<ValidationCode>` type in Rust.
```suggestion
	// New validation code.
	NewValidationCode *ValidationCode `scale:"3"`
```